src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java

```diff
@@ -0,0 +1,297 @@
+        Pageable pageable = PageRequest.of(0, 10);
+        Page<Book> emptyBookPage = new PageImpl<>(Collections.emptyList());
+
+        Mockito.when(bookRepository.findAll(pageable)).thenReturn(emptyBookPage);
+
+        Exception exception = assertThrows(EntityNotFoundException.class, () -> {
+            bookService.findAll(pageable);
+        });
+
+        String expectedMessage = "Can't find any books";
+        String actualMessage = exception.getMessage();
+
+        assertTrue(actualMessage.contains(expectedMessage));
+    }
+
+    @Test
+    @DisplayName("Create new Book")
+    @Sql(scripts = {
+            "classpath:database/book/remove-all-from-books-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void createNewBook_ValidData_OK() {
+        Set<Long> categoryIds = new HashSet<>();
+        categoryIds.add(1L);
+
+        BookRequestDto dto = new BookRequestDto();
+        dto.setTitle("Lenore")
+                .setAuthor("EdgarAllanPoe")
+                .setIsbn("9780306406157")
+                .setPrice(BigDecimal.valueOf(10.50))
+                .setDescription("test")
+                .setCoverImage("test@test.com")
+                .setCategoryIds(categoryIds);
+
+        Book book = new Book();
+        book.setTitle("Lenore")
+                .setAuthor("EdgarAllanPoe")
+                .setIsbn("9780306406157")
+                .setPrice(BigDecimal.valueOf(10.50))
+                .setDescription("test")
+                .setCoverImage("test@test.com");
+
+        List<Category> categories = categoryIds.stream()
+                .map(id -> {
+                    Category category = new Category();
+                    category.setId(id);
+                    return category;
+                })
+                .collect(Collectors.toList());
+
+        BookDtoWithoutCategoriesIds bookDtoWithoutCategories = new BookDtoWithoutCategoriesIds();
+        bookDtoWithoutCategories.setTitle("Lenore");
+
+        BookResponseDto bookResponseDto = new BookResponseDto();
+        bookResponseDto.setTitle("Lenore");
+
+        Mockito.when(categoryRepository.findAllById(categoryIds)).thenReturn(categories);
+        Mockito.when(bookMapper.toModel(dto)).thenReturn(book);
+        Mockito.when(bookRepository.save(book)).thenReturn(book);
+        Mockito.when(bookMapper.toDto(book)).thenReturn(bookResponseDto);
+
+        BookResponseDto savedBookDto = bookService.save(dto);
+
+        Mockito.verify(bookRepository).save(book);
+        Set<Category> categorySet = new HashSet<>(categories);
+        assertEquals(categorySet, book.getCategories());
+
+        assertNotNull(savedBookDto);
+        assertEquals(dto.getTitle(), savedBookDto.getTitle());
+    }
+
+    @Test
+    @DisplayName("Update existing Book")
+    @Sql(scripts = {
+            "classpath:database/book/remove-all-from-books-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)

```

Comment:
Services tests should be unit, so we need to mock repository behavior. And test repo separatelly